lib/Dialect/FIRRTL/Transforms/LowerAnnotations.cpp

```diff
@@ -474,6 +476,229 @@ LogicalResult LowerAnnotationsPass::applyAnnotation(DictionaryAttr anno,
+  // to be made per-module.
+  LLVM_DEBUG({ llvm::dbgs() << "Analyzing wiring problems:\n"; });
+  DenseMap<FModuleLike, ModuleModifications> moduleModifications;
+  for (auto [index, problem] : llvm::enumerate(state.wiringProblems)) {
+    // This is a unique index that is assigned to this specific wiring problem
+    // and is used as a key during wiring to know which Values (ports, sources,
+    // or sinks) should be connected.
+    auto source = problem.source;
+    auto sink = problem.sink;
+
+    // If the source and sink are in the same module, just wire them up.
+    if (sink.getParentBlock() == source.getParentBlock()) {
+      auto builder = ImplicitLocOpBuilder::atBlockEnd(UnknownLoc::get(context),
+                                                      sink.getParentBlock());
+      builder.create<ConnectOp>(sink, source);
+      continue;
+    }
+
+    // Pre-populate source/sink module modifications connection values.
+    FModuleLike sourceModule = getModule(source);
+    FModuleLike sinkModule = getModule(sink);
+
+    if (isa<FExtModuleOp>(sourceModule) || isa<FExtModuleOp>(sinkModule)) {
+      auto diag = mlir::emitError(source.getLoc())
+                  << "This source is involved with a Wiring Problem which "
+                     "includes an External Module port and External Module "
+                     "ports anre not supported.";
+      diag.attachNote(sink.getLoc()) << "The sink is here.";
+      return failure();
+    }
+
+    auto sourcePaths = state.instancePathCache.getAbsolutePaths(
+        cast<hw::HWModuleLike>(*sourceModule));
+    auto sinkPaths = state.instancePathCache.getAbsolutePaths(
+        cast<hw::HWModuleLike>(*sinkModule));
+
+    if (sourcePaths.size() != 1 || sinkPaths.size() != 1) {
+      auto diag =
+          mlir::emitError(source.getLoc())
+          << "This source is involved with a Wiring Problem where the source "
+             "or the sink are multiply instantiated and this is not supported.";
+      diag.attachNote(sink.getLoc()) << "The sink is here.";
+      return failure();
+    }
+
+    FModuleOp lca =
+        cast<FModuleOp>(instanceGraph.getTopLevelNode()->getModule());
+    auto sources = sourcePaths[0];
+    auto sinks = sinkPaths[0];
+    while (!sources.empty() && !sinks.empty()) {
+      if (sources[0] != sinks[0])
+        break;
+      auto newLCA = sources[0];
+      lca = cast<FModuleOp>(newLCA.getReferencedModule());
+      sources = sources.drop_front();
+      sinks = sinks.drop_front();
+    }
+
+    // Pre-populate the connectionMap of the module with the source and sink.
+    moduleModifications[sourceModule].connectionMap[index] = source;
+    moduleModifications[sinkModule].connectionMap[index] = sink;
+
+    // Record ports that should be added to each module along the LCA path.
+    auto tpe = RefType::get(cast<FIRRTLBaseType>(problem.sink.getType()));
+    for (auto source : sources) {
+      auto mod = cast<FModuleOp>(source.getReferencedModule());
```

Comment:
Maybe Query the `state.instancegraph` for `getReferencedModule` ?

Reply:
Great catch and ditto for below. This thing is such a foot-gun and even though I know not to use it, I still forget. 🙈